App Crash on Share QR Code

**Steps to reproduce:**

1. Launch App.
2. Goto Settings.
3. Tap `Import/Export Settings via QR Code` button.
4. Tap share button.

```java
E/com.onaio.steps.helper.FileUtil: java.io.FileNotFoundException: /storage/emulated/0/steps/settings/.steps-settings-hash: open failed: EPERM (Operation not permitted)
        at libcore.io.IoBridge.open(IoBridge.java:492)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:236)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:186)
        at com.onaio.steps.helper.FileUtil.write(FileUtil.java:163)
        at com.onaio.steps.utils.QRCodeUtils.saveToDisk(QRCodeUtils.java:183)
        at com.onaio.steps.tasks.SaveQRCodeAsyncTask.doInBackground(SaveQRCodeAsyncTask.java:33)
        at com.onaio.steps.tasks.SaveQRCodeAsyncTask.doInBackground(SaveQRCodeAsyncTask.java:17)
        at android.os.AsyncTask$3.call(AsyncTask.java:394)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:305)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:923)
     Caused by: android.system.ErrnoException: open failed: EPERM (Operation not permitted)
        at libcore.io.Linux.open(Native Method)
        at libcore.io.ForwardingOs.open(ForwardingOs.java:166)
        at libcore.io.BlockGuardOs.open(BlockGuardOs.java:254)
        at libcore.io.ForwardingOs.open(ForwardingOs.java:166)
        at android.app.ActivityThread$AndroidOs.open(ActivityThread.java:7542)
        at libcore.io.IoBridge.open(IoBridge.java:478)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:236) 
        at java.io.FileOutputStream.<init>(FileOutputStream.java:186) 
        at com.onaio.steps.helper.FileUtil.write(FileUtil.java:163) 
        at com.onaio.steps.utils.QRCodeUtils.saveToDisk(QRCodeUtils.java:183) 
        at com.onaio.steps.tasks.SaveQRCodeAsyncTask.doInBackground(SaveQRCodeAsyncTask.java:33) 
        at com.onaio.steps.tasks.SaveQRCodeAsyncTask.doInBackground(SaveQRCodeAsyncTask.java:17) 
        at android.os.AsyncTask$3.call(AsyncTask.java:394) 
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) 
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:305) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:923) 
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.onaio.steps, PID: 27094
    android.os.FileUriExposedException: file:///storage/emulated/0/Pictures/my-steps-settings.png exposed beyond app through ClipData.Item.getUri()
        at android.os.StrictMode.onFileUriExposed(StrictMode.java:2141)
        at android.net.Uri.checkFileUriExposed(Uri.java:2391)
        at android.content.ClipData.prepareToLeaveProcess(ClipData.java:964)
        at android.content.Intent.prepareToLeaveProcess(Intent.java:11119)
        at android.content.Intent.prepareToLeaveProcess(Intent.java:11125)
        at android.content.Intent.prepareToLeaveProcess(Intent.java:11104)
        at android.app.Instrumentation.execStartActivity(Instrumentation.java:1722)
        at android.app.Activity.startActivityForResult(Activity.java:5320)
        at androidx.activity.ComponentActivity.startActivityForResult(ComponentActivity.java:597)
        at android.app.Activity.startActivityForResult(Activity.java:5278)
        at androidx.activity.ComponentActivity.startActivityForResult(ComponentActivity.java:583)
        at android.app.Activity.startActivity(Activity.java:5664)
        at android.app.Activity.startActivity(Activity.java:5617)
        at com.onaio.steps.handler.actions.ShareHandler$1.onSuccessfulSave(ShareHandler.java:60)
        at com.onaio.steps.tasks.SaveQRCodeAsyncTask.onPostExecute(SaveQRCodeAsyncTask.java:48)
        at com.onaio.steps.tasks.SaveQRCodeAsyncTask.onPostExecute(SaveQRCodeAsyncTask.java:17)
        at android.os.AsyncTask.finish(AsyncTask.java:771)
        at android.os.AsyncTask.access$900(AsyncTask.java:199)
        at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:788)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

--- a/src/main/AndroidManifest.xml
+++ b/src/main/AndroidManifest.xml
@@ -103,9 +103,23 @@
             </intent-filter>
         </activity>
 
+        <provider
+            android:authorities="${applicationId}"
+            android:name="androidx.core.content.FileProvider"
+            android:exported="false"
+            android:grantUriPermissions="true">
+
+            <meta-data
+                android:name="android.support.FILE_PROVIDER_PATHS"
+                android:resource="@xml/file_provider_paths" />
+
+        </provider>
+        
     </application>
 
     <queries>
         <package android:name="org.odk.collect.android" />
     </queries>
+    
+    
 </manifest>

--- a/src/main/java/com/onaio/steps/handler/actions/ShareHandler.java
+++ b/src/main/java/com/onaio/steps/handler/actions/ShareHandler.java
@@ -55,9 +55,9 @@ public class ShareHandler implements IMenuHandler, IMenuPreparer {
             public void onSuccessfulSave() {
                 Intent intent = new Intent(Intent.ACTION_SEND);
                 intent.setType("image/*");
-                intent.putExtra(Intent.EXTRA_STREAM, Uri.parse("file://" + QRCodeUtils.QR_CODE_FILEPATH));
+                //intent.putExtra(Intent.EXTRA_STREAM, Uri.parse("file://" + QRCodeUtils.QR_CODE_FILEPATH));
 
-                settingsImportExportActivity.startActivity(Intent.createChooser(intent, settingsImportExportActivity.getString(R.string.share_qr_code_title)));
+                //settingsImportExportActivity.startActivity(Intent.createChooser(intent, settingsImportExportActivity.getString(R.string.share_qr_code_title)));
             }
 
             @Override

--- a/src/main/java/com/onaio/steps/handler/actions/ShareHandler.java
+++ b/src/main/java/com/onaio/steps/handler/actions/ShareHandler.java
@@ -34,15 +34,15 @@ import com.onaio.steps.utils.ViewUtils;
 
 import java.io.File;
 
-public class ShareHandler implements IMenuHandler, IMenuPreparer {
+public class ShareHandler implements IMenuHandler, IMenuPreparer, QRBitmapSaveListener {
 
-    private final SettingsImportExportActivity settingsImportExportActivity;
+    private final SettingsImportExportActivity activity;
     private final boolean qrDisplayed;
     private static final int MENU_ID = R.id.menu_item_settings_share;
     private Menu menu;
 
     public ShareHandler(SettingsImportExportActivity settingsImportExportActivity, boolean qrDisplayed) {
-        this.settingsImportExportActivity = settingsImportExportActivity;
+        this.activity = settingsImportExportActivity;
         this.qrDisplayed = qrDisplayed;
     }
 
@@ -53,27 +53,7 @@ public class ShareHandler implements IMenuHandler, IMenuPreparer {
 
     @Override
     public boolean open() {
-        SaveQRCodeAsyncTask saveQRCodeAsyncTask = new SaveQRCodeAsyncTask(settingsImportExportActivity, settingsImportExportActivity.getQrCodeBitmap(), new QRBitmapSaveListener() {
-            @Override
-            public void onSuccessfulSave() {
-                Intent intent = new Intent(Intent.ACTION_SEND);
-                intent.setType("image/*");
-                intent.putExtra(Intent.EXTRA_STREAM,
-                        FileProvider.getUriForFile(settingsImportExportActivity,
-                                BuildConfig.APPLICATION_ID,
-                                new File(settingsImportExportActivity.getFilesDir() + File.separator, QRCodeUtils.QR_CODE_FILEPATH)
-                        )
-                );
-
-                settingsImportExportActivity.startActivity(Intent.createChooser(intent, settingsImportExportActivity.getString(R.string.share_qr_code_title)));
-            }
-
-            @Override
-            public void onError(Exception e) {
-                ViewUtils.showCustomToast(settingsImportExportActivity, settingsImportExportActivity.getString(R.string.qr_code_share_error));
-            }
-        });
-
+        SaveQRCodeAsyncTask saveQRCodeAsyncTask = new SaveQRCodeAsyncTask(activity, activity.getQrCodeBitmap(), this);
         saveQRCodeAsyncTask.execute();
         return true;
     }
@@ -100,5 +80,24 @@ public class ShareHandler implements IMenuHandler, IMenuPreparer {
         MenuItem item = menu.findItem(MENU_ID);
         item.setVisible(true);
         item.setEnabled(true);
+    }
+
+    @Override
+    public void onSuccessfulSave() {
+        Intent intent = new Intent(Intent.ACTION_SEND);
+        intent.setType("image/*");
+        intent.putExtra(Intent.EXTRA_STREAM,
+                FileProvider.getUriForFile(activity,
+                        BuildConfig.APPLICATION_ID,
+                        new File(activity.getFilesDir() + File.separator, QRCodeUtils.QR_CODE_FILEPATH)
+                )
+        );
+
+        activity.startActivity(Intent.createChooser(intent, activity.getString(R.string.share_qr_code_title)));
+    }
+
+    @Override
+    public void onError(Exception e) {
+        ViewUtils.showCustomToast(activity, activity.getString(R.string.qr_code_share_error));
     }
 }

--- a/src/main/java/com/onaio/steps/handler/actions/ShareHandler.java
+++ b/src/main/java/com/onaio/steps/handler/actions/ShareHandler.java
@@ -17,11 +17,12 @@
 package com.onaio.steps.handler.actions;
 
 import android.content.Intent;
-import android.net.Uri;
 import android.view.Menu;
 import android.view.MenuItem;
-import android.view.View;
 
+import androidx.core.content.FileProvider;
+
+import com.onaio.steps.BuildConfig;
 import com.onaio.steps.R;
 import com.onaio.steps.activities.SettingsImportExportActivity;
 import com.onaio.steps.handler.interfaces.IMenuHandler;
@@ -31,10 +32,12 @@ import com.onaio.steps.tasks.SaveQRCodeAsyncTask;
 import com.onaio.steps.utils.QRCodeUtils;
 import com.onaio.steps.utils.ViewUtils;
 
+import java.io.File;
+
 public class ShareHandler implements IMenuHandler, IMenuPreparer {
 
-    private SettingsImportExportActivity settingsImportExportActivity;
-    private boolean qrDisplayed;
+    private final SettingsImportExportActivity settingsImportExportActivity;
+    private final boolean qrDisplayed;
     private static final int MENU_ID = R.id.menu_item_settings_share;
     private Menu menu;
 
@@ -55,9 +58,14 @@ public class ShareHandler implements IMenuHandler, IMenuPreparer {
             public void onSuccessfulSave() {
                 Intent intent = new Intent(Intent.ACTION_SEND);
                 intent.setType("image/*");
-                //intent.putExtra(Intent.EXTRA_STREAM, Uri.parse("file://" + QRCodeUtils.QR_CODE_FILEPATH));
+                intent.putExtra(Intent.EXTRA_STREAM,
+                        FileProvider.getUriForFile(settingsImportExportActivity,
+                                BuildConfig.APPLICATION_ID,
+                                new File(settingsImportExportActivity.getFilesDir() + File.separator, QRCodeUtils.QR_CODE_FILEPATH)
+                        )
+                );
 
-                //settingsImportExportActivity.startActivity(Intent.createChooser(intent, settingsImportExportActivity.getString(R.string.share_qr_code_title)));
+                settingsImportExportActivity.startActivity(Intent.createChooser(intent, settingsImportExportActivity.getString(R.string.share_qr_code_title)));
             }
 
             @Override

--- a/src/main/java/com/onaio/steps/helper/Constants.java
+++ b/src/main/java/com/onaio/steps/helper/Constants.java
@@ -16,10 +16,6 @@
 
 package com.onaio.steps.helper;
 
-import android.os.Environment;
-
-import java.io.File;
-
 public class Constants {
 
     public static final String APP_DIR = "STEPS";
@@ -78,9 +74,6 @@ public class Constants {
     public static final String SETTINGS_AUTH_TIME = "settings_auth_time";
 
     public static final long SETTINGS_AUTH_TIMEOUT = 60000;//the number of milliseconds after authentication in the settings screen before authentication can be asked
-
-    public static final String STEPS_ROOT_FOLDER = Environment.getExternalStorageDirectory()
-            + File.separator + "steps";
 
     public static final String SETTINGS = "settings";
 

--- a/src/main/java/com/onaio/steps/helper/Constants.java
+++ b/src/main/java/com/onaio/steps/helper/Constants.java
@@ -82,7 +82,7 @@ public class Constants {
     public static final String STEPS_ROOT_FOLDER = Environment.getExternalStorageDirectory()
             + File.separator + "steps";
 
-    public static final String SETTINGS = STEPS_ROOT_FOLDER + File.separator + "settings";
+    public static final String SETTINGS = "settings";
 
     public static final String UNIQUE_DEVICE_ID = "unique_device_id";
 }

--- a/src/main/java/com/onaio/steps/utils/QRCodeUtils.java
+++ b/src/main/java/com/onaio/steps/utils/QRCodeUtils.java
@@ -67,7 +67,7 @@ public class QRCodeUtils {
 
     private static final String TAG = QRCodeUtils.class.getName();
 
-    public static final String QR_CODE_FILEPATH = Environment.getExternalStorageDirectory() + File.separator + Environment.DIRECTORY_PICTURES + File.separator + "my-steps-settings.png";
+    public static final String QR_CODE_FILEPATH = Constants.SETTINGS + File.separator + "my-steps-settings.png";
     private static final int QR_CODE_SIDE_LENGTH = 400; // in pixels
     private static final String SETTINGS_MD5_FILE = ".steps-settings-hash";
     static final String MD5_CACHE_PATH = Constants.SETTINGS + File.separator + SETTINGS_MD5_FILE;
@@ -132,15 +132,17 @@ public class QRCodeUtils {
 
         Bitmap bitmap = null;
 
+        String basePath = activity.getFilesDir().getAbsolutePath();
+
         // check if settings directory exists, if not then create one
-        File writeDir = new File(Constants.SETTINGS);
+        File writeDir = new File(basePath + File.separator + Constants.SETTINGS);
         if (!writeDir.exists()) {
             if (!writeDir.mkdirs()) {
                 Log.e(TAG, "Error creating directory " + writeDir.getAbsolutePath());
             }
         }
 
-        File mdCacheFile = new File(MD5_CACHE_PATH);
+        File mdCacheFile = new File(basePath + File.separator + MD5_CACHE_PATH);
         if (mdCacheFile.exists()) {
             byte[] cachedMessageDigest = FileUtil.read(mdCacheFile);
 
@@ -167,21 +169,21 @@ public class QRCodeUtils {
 
     public static void saveToDisk(AppCompatActivity activity, Bitmap bitmap) throws JSONException, NoSuchAlgorithmException {
 
-        String settingsJSON = exportSettingsToJSON(activity);
+        //String settingsJSON = exportSettingsToJSON(activity);
 
-        MessageDigest md = MessageDigest.getInstance("MD5");
-        md.update(settingsJSON.getBytes());
-        byte[] messageDigest = md.digest();
+        //MessageDigest md = MessageDigest.getInstance("MD5");
+        //md.update(settingsJSON.getBytes());
+        //byte[] messageDigest = md.digest();
 
-        File mdCacheFile = new File(MD5_CACHE_PATH);
-
+        //File mdCacheFile = new File(MD5_CACHE_PATH);
+        String basePath = activity.getFilesDir().getAbsolutePath();
         if (bitmap != null) {
             // Save the QRCode to disk
-                logInfo("Saving QR Code to disk... : " + QR_CODE_FILEPATH);
-                FileUtil.saveBitmapToFile(bitmap, QR_CODE_FILEPATH);
+                logInfo("Saving QR Code to disk... : " + basePath + File.separator + QR_CODE_FILEPATH);
+                FileUtil.saveBitmapToFile(bitmap, basePath + File.separator + QR_CODE_FILEPATH);
 
-                FileUtil.write(mdCacheFile, messageDigest);
-                logInfo("Updated %s file contents", SETTINGS_MD5_FILE);
+                /*FileUtil.write(mdCacheFile, messageDigest);
+                logInfo("Updated %s file contents", SETTINGS_MD5_FILE);*/
         }
     }
 

--- a/src/main/java/com/onaio/steps/utils/QRCodeUtils.java
+++ b/src/main/java/com/onaio/steps/utils/QRCodeUtils.java
@@ -18,7 +18,6 @@ import android.content.Context;
 import android.graphics.Bitmap;
 import android.graphics.BitmapFactory;
 import android.graphics.Color;
-import android.os.Environment;
 import android.util.Log;
 
 import androidx.annotation.NonNull;
@@ -154,7 +153,7 @@ public class QRCodeUtils {
                 logInfo("Loading QRCode from the disk...");
                 BitmapFactory.Options options = new BitmapFactory.Options();
                 options.inPreferredConfig = Bitmap.Config.ARGB_8888;
-                bitmap = FileUtil.getBitmap(QR_CODE_FILEPATH, options);
+                bitmap = FileUtil.getBitmap(basePath + File.separator + QR_CODE_FILEPATH, options);
             }
         }
 
@@ -169,21 +168,21 @@ public class QRCodeUtils {
 
     public static void saveToDisk(AppCompatActivity activity, Bitmap bitmap) throws JSONException, NoSuchAlgorithmException {
 
-        //String settingsJSON = exportSettingsToJSON(activity);
+        String settingsJSON = exportSettingsToJSON(activity);
 
-        //MessageDigest md = MessageDigest.getInstance("MD5");
-        //md.update(settingsJSON.getBytes());
-        //byte[] messageDigest = md.digest();
+        MessageDigest md = MessageDigest.getInstance("MD5");
+        md.update(settingsJSON.getBytes());
+        byte[] messageDigest = md.digest();
 
-        //File mdCacheFile = new File(MD5_CACHE_PATH);
         String basePath = activity.getFilesDir().getAbsolutePath();
+        File mdCacheFile = new File(basePath + File.separator + MD5_CACHE_PATH);
         if (bitmap != null) {
             // Save the QRCode to disk
                 logInfo("Saving QR Code to disk... : " + basePath + File.separator + QR_CODE_FILEPATH);
                 FileUtil.saveBitmapToFile(bitmap, basePath + File.separator + QR_CODE_FILEPATH);
 
-                /*FileUtil.write(mdCacheFile, messageDigest);
-                logInfo("Updated %s file contents", SETTINGS_MD5_FILE);*/
+                FileUtil.write(mdCacheFile, messageDigest);
+                logInfo("Updated %s file contents", SETTINGS_MD5_FILE);
         }
     }
 

--- a/src/main/res/xml/file_provider_paths.xml
+++ b/src/main/res/xml/file_provider_paths.xml
@@ -1,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<paths>
+    <files-path name="settings" path="settings" />
+</paths>

--- a/src/test/java/com/onaio/steps/handler/actions/ShareHandlerTest.java
+++ b/src/test/java/com/onaio/steps/handler/actions/ShareHandlerTest.java
@@ -16,8 +16,19 @@
 
 package com.onaio.steps.handler.actions;
 
+import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import android.content.Intent;
+import android.view.Menu;
+import android.view.MenuItem;
 
 import com.onaio.steps.R;
 import com.onaio.steps.StepsTestRunner;
@@ -25,20 +36,22 @@ import com.onaio.steps.shadows.TestSettingsImportExportActivity;
 
 import org.junit.Before;
 import org.junit.Test;
+import org.mockito.ArgumentCaptor;
 import org.robolectric.Robolectric;
 
 public class ShareHandlerTest extends StepsTestRunner {
 
-    private TestSettingsImportExportActivity settingsImportExportActivity;
+    private TestSettingsImportExportActivity activity;
     private ShareHandler shareHandler;
+    private static final int MENU_ID = R.id.menu_item_settings_share;
 
     @Before
     public void setUp() {
-        settingsImportExportActivity = Robolectric.buildActivity(TestSettingsImportExportActivity.class)
+        activity = spy(Robolectric.buildActivity(TestSettingsImportExportActivity.class)
                 .create()
                 .visible()
-                .get();
-        shareHandler = new ShareHandler(settingsImportExportActivity, false);
+                .get());
+        shareHandler = new ShareHandler(activity, false);
     }
 
     @Test
@@ -51,9 +64,33 @@ public class ShareHandlerTest extends StepsTestRunner {
         assertFalse(shareHandler.shouldOpen(R.id.importCodeImageBtn));
     }
 
-
     @Test
     public void shouldDeactivate() {
         assertTrue(shareHandler.shouldDeactivate());
+    }
+
+    @Test
+    public void testActivateShouldVisibleAndEnableMenuItem() {
+        Menu menu = mock(Menu.class);
+        MenuItem menuItem = mock(MenuItem.class);
+
+        when(menu.findItem(MENU_ID)).thenReturn(menuItem);
+
+        shareHandler.withMenu(menu);
+        shareHandler.activate();
+
+        verify(menuItem, times(1)).setVisible(eq(true));
+        verify(menuItem, times(1)).setEnabled(eq(true));
+    }
+
+    @Test
+    public void testOnSuccessfulSaveShouldStartSharingIntent() {
+        shareHandler.onSuccessfulSave();
+
+        ArgumentCaptor<Intent> intentArgumentCaptor = ArgumentCaptor.forClass(Intent.class);
+        verify(activity, times(1)).startActivity(intentArgumentCaptor.capture());
+
+        Intent intent = intentArgumentCaptor.getValue();
+        assertEquals("android.intent.action.CHOOSER", intent.getAction());
     }
 }

